From af591bf68b568cce547e5c30617e31555052d401 Mon Sep 17 00:00:00 2001
From: Duy Truong <duytruong@google.com>
Date: Mon, 27 Apr 2026 15:47:29 -0700
Subject: [PATCH 19/27] Fix security check bypass

Ensure that security checks (mime type and size limits) are applied
in all image resolution paths. Previously, the checks were only in
onHeaderDecoded, which was bypassed in some resolveImage variants.

This was autogenerated by MendIt (go/androidmendit)

Bug: 501447778
Test: atest LocalImageResolverTest
Flag: EXEMPT BUGFIX
Change-Id: Ief9e76d49f069b42568d1037452b1fbfa508d453
(cherry picked from commit ea59bf7683410de0cb9fe57ff3c56bd5b39d7d7e)
---
 .../internal/widget/LocalImageResolver.java   | 25 ++++++++++++-------
 .../widget/LocalImageResolverTest.java        | 16 ++++++++++++
 2 files changed, 32 insertions(+), 9 deletions(-)

diff --git a/core/java/com/android/internal/widget/LocalImageResolver.java b/core/java/com/android/internal/widget/LocalImageResolver.java
index 97e66d9ddfd0..71dbbe06ba73 100644
--- a/core/java/com/android/internal/widget/LocalImageResolver.java
+++ b/core/java/com/android/internal/widget/LocalImageResolver.java
@@ -220,6 +220,7 @@ public class LocalImageResolver {
     private static Drawable resolveImage(ImageDecoder.Source source, int maxWidth, int maxHeight) {
         try {
             return ImageDecoder.decodeDrawable(source, (decoder, info, unused) -> {
+                checkHeader(info);
                 if (maxWidth <= 0 || maxHeight <= 0) {
                     return;
                 }
@@ -244,8 +245,9 @@ public class LocalImageResolver {
             });
 
         // ImageDecoder documentation is misleading a bit - it'll throw NotFoundException
-        // in some cases despite it not saying so.
-        } catch (IOException | Resources.NotFoundException e) {
+        // in some cases despite it not saying so. We also throw a RuntimeException if the
+        // image size is unreasonable (see checkHeader).
+        } catch (IOException | RuntimeException e) {
             Log.d(TAG, "Couldn't use ImageDecoder for drawable, falling back to non-resized load.");
             return null;
         }
@@ -258,6 +260,18 @@ public class LocalImageResolver {
 
     private static void onHeaderDecoded(ImageDecoder decoder, ImageDecoder.ImageInfo info,
             int maxWidth, int maxHeight) {
+        checkHeader(info);
+        final Size size = info.getSize();
+
+        final int originalSize = Math.max(size.getHeight(), size.getWidth());
+        final int maxSize = Math.max(maxWidth, maxHeight);
+        final double ratio = (originalSize > maxSize)
+                ? originalSize * 1f / maxSize
+                : 1.0;
+        decoder.setTargetSampleSize(getPowerOfTwoForSampleRatio(ratio));
+    }
+
+    private static void checkHeader(ImageDecoder.ImageInfo info) {
         final Size size = info.getSize();
 
         final String mimeType = info.getMimeType();
@@ -290,13 +304,6 @@ public class LocalImageResolver {
                     "Image dimensions (" + size.getWidth() + "x" + size.getHeight()
                             + ") exceed the maximum allowed size.");
         }
-
-        final int originalSize = Math.max(size.getHeight(), size.getWidth());
-        final int maxSize = Math.max(maxWidth, maxHeight);
-        final double ratio = (originalSize > maxSize)
-                ? originalSize * 1f / maxSize
-                : 1.0;
-        decoder.setTargetSampleSize(getPowerOfTwoForSampleRatio(ratio));
     }
 
     /**
diff --git a/core/tests/coretests/src/com/android/internal/widget/LocalImageResolverTest.java b/core/tests/coretests/src/com/android/internal/widget/LocalImageResolverTest.java
index 474e3c3f010b..31acda8b2bf6 100644
--- a/core/tests/coretests/src/com/android/internal/widget/LocalImageResolverTest.java
+++ b/core/tests/coretests/src/com/android/internal/widget/LocalImageResolverTest.java
@@ -290,6 +290,14 @@ public class LocalImageResolverTest {
         LocalImageResolver.resolveImage(uri, mContext);
     }
 
+    @Test(expected = RuntimeException.class)
+    public void resolveImage_withSize_veryLargeResource_throwsException() {
+        // Passing in an unreasonably large image should throw an exception.
+        Uri uri = Uri.parse("android.resource://"
+                + mContext.getPackageName() + "/" + R.drawable.test16000x16000);
+        LocalImageResolver.resolveImage(uri, mContext, 480, 480);
+    }
+
     @Test
     public void resolveResourcesForIcon_notAResourceIcon_returnsNull() {
         Icon icon = Icon.createWithContentUri(Uri.parse("some_uri"));
@@ -342,4 +350,12 @@ public class LocalImageResolverTest {
                 + mContext.getPackageName() + "/" + R.raw.dng_opcode_MapTable_ProcessArea);
         LocalImageResolver.resolveImage(uri, mContext);
     }
+
+    @Test(expected = RuntimeException.class)
+    public void resolveImage_asset_invalidMimeType_withSize() {
+        // dng mimetype is not supported
+        Uri uri = Uri.parse("android.resource://"
+                + mContext.getPackageName() + "/" + R.raw.dng_opcode_MapTable_ProcessArea);
+        LocalImageResolver.resolveImage(uri, mContext, 480, 480);
+    }
 }
-- 
2.55.0

